sui_core/src/checkpoints/reconstruction.rs

```diff
@@ -0,0 +1,152 @@
+            // Otherwise ignore the link.
+            let n1 = &frag.proposer.0.authority;
+            if *proposals
+                .entry(*n1)
+                .or_insert_with(|| frag.proposer.0.checkpoint.clone())
+                != frag.proposer.0.checkpoint
+            {
+                continue;
+            }
+
+            let n2 = &frag.other.0.authority;
+            if *proposals
+                .entry(*n2)
+                .or_insert_with(|| frag.other.0.checkpoint.clone())
+                != frag.other.0.checkpoint
+            {
+                continue;
+            }
+
+            // Add to the links we will consider.
+            fragments_used.push(frag);
+
+            // Merge the link.
+            let (top, weight) = span.merge(n1, n2);
+
+            // We have found a connected component larger than the 2/3 threshold
+            if weight >= committee.quorum_threshold() {
+                // Get all links that are part of this component
+                let mut active_links: VecDeque<_> = fragments_used
+                    .into_iter()
+                    .filter(|frag| span.top_node(&frag.proposer.0.authority).0 == top)
+                    .collect();
+
+                let mut global = GlobalCheckpoint::new();
+                while let Some(link) = active_links.pop_front() {
+                    match global.insert(link.diff.clone()) {
+                        Ok(_) | Err(WaypointError::NothingToDo) => {} // Do nothing
+                        Err(WaypointError::CannotConnect) => {
+                            // Reinsert the fragment at the end
+                            active_links.push_back(link);
+                        }
+                        other => {
+                            // This is bad news, we did not intend to fail here.
+                            // We should have checked all conditions to avoid being
+                            // in this situation. TODO: audit this.
+                            println!("Unexpected result: {:?}", other);
+                            unreachable!();
+                        }
+                    }
+                }
+
+                return Ok(Some(FragmentReconstruction { global, committee }));
+            }
+        }
+
+        // If we run out of candidates with no checkpoint, there is no
+        // checkpoint yet.
+        Ok(None)
+    }
+}
+
+// A structure that stores a set of spanning trees, and that supports addition
+// of links to merge them, and construct ever growing components.
+struct SpanGraph {
+    nodes: HashMap<AuthorityName, (AuthorityName, usize)>,
+}
+
+impl SpanGraph {
+    /// Initialize the graph with each authority just pointing to itself.
+    pub fn new(committee: &Committee) -> SpanGraph {
+        let mut nodes: HashMap<AuthorityName, (AuthorityName, usize)> = HashMap::new();
+        for (n, w) in &committee.voting_rights {
+            nodes.insert(*n, (*n, *w));
+        }
+        SpanGraph { nodes }
```

Comment:
You could use the implementation of `FromIterator` for `HashMap`.

Reply:
Excellent idea!